sha.go

```diff
@@ -559,6 +587,214 @@ func (h *sha512Hash) UnmarshalBinary(b []byte) error {
 	return nil
 }
 
+// NewSHA3_224 returns a new SHA3-224 hash.
+func NewSHA3_224() hash.Hash {
+	return &sha3_224Hash{
+		evpHash: newEvpHash(crypto.SHA3_224, 224/8, 64),
+	}
+}
+
+type sha3_224Hash struct {
+	*evpHash
+	out [224 / 8]byte
+}
+
+func (h *sha3_224Hash) Sum(in []byte) []byte {
+	h.sum(h.out[:])
+	return append(in, h.out[:]...)
+}
+
+// NewSHA3_256 returns a new SHA3-256 hash.
+func NewSHA3_256() hash.Hash {
+	return &sha3_256Hash{
+		evpHash: newEvpHash(crypto.SHA3_256, 256/8, 64),
+	}
+}
+
+type sha3_256Hash struct {
+	*evpHash
+	out [256 / 8]byte
+}
+
+func (h *sha3_256Hash) Sum(in []byte) []byte {
+	h.sum(h.out[:])
+	return append(in, h.out[:]...)
+}
+
+// NewSHA3_384 returns a new SHA3-384 hash.
+func NewSHA3_384() hash.Hash {
+	return &sha3_384Hash{
+		evpHash: newEvpHash(crypto.SHA3_384, 384/8, 128),
+	}
+}
+
+type sha3_384Hash struct {
+	*evpHash
+	out [384 / 8]byte
+}
+
+func (h *sha3_384Hash) Sum(in []byte) []byte {
+	h.sum(h.out[:])
+	return append(in, h.out[:]...)
+}
+
+// NewSHA3_512 returns a new SHA3-512 hash.
+func NewSHA3_512() hash.Hash {
+	return &sha3_512Hash{
+		evpHash: newEvpHash(crypto.SHA3_512, 512/8, 128),
+	}
+}
+
+type sha3_512Hash struct {
+	*evpHash
+	out [512 / 8]byte
+}
+
+func (h *sha3_512Hash) Sum(in []byte) []byte {
+	h.sum(h.out[:])
+	return append(in, h.out[:]...)
+}
+
+// keccakState layout is taken from
+// https://github.com/openssl/openssl/blob/9522f0a6a9c49d650c773f089ed84b0c1ee0368b/include/internal/sha3.h#L34
+type keccakState struct {
+	A [25]uint64
+	// Assumes 64-bit platform, as sadly block_size/md_size/bufsz
+	// are declared as size_t
+	block_size, md_size, bufsz uint64
+	x                          [1600/8 - 32 + 1]byte
+}
+
+const (
+	sha3Magic224        = "sha3\x01"
+	sha3Magic256        = "sha3\x02"
+	sha3Magic384        = "sha3\x03"
+	sha3Magic512        = "sha3\x04"
+	keccakMarshaledSize = len(sha3Magic512) + 25*8 + 3*8 + 168 + 1
+)
+
+func marshalkeccak(magic string, d *keccakState) []byte {
+	b := make([]byte, 0, keccakMarshaledSize)
+	b = append(b, magic...)
+	for i := 0; i < len(d.A); i++ {
+		b = appendUint64(b, d.A[i])
+	}
+	// Maybe not portable across 32/64 bit platforms
+	b = appendUint64(b, d.block_size)
+	b = appendUint64(b, d.md_size)
+	b = appendUint64(b, d.bufsz)
+	// Maybe not portable accross be/le platforms
+	b = append(b, d.x[:]...)
+	// Honestly, only use this across same platforms
+	return b
+}
+
+func unmarshalkeccak(d *keccakState, b []byte) {
+	for i := 0; i < len(d.A); i++ {
+		b, d.A[i] = consumeUint64(b)
+	}
+	b, d.block_size = consumeUint64(b)
+	b, d.md_size = consumeUint64(b)
+	b, d.bufsz = consumeUint64(b)
+	b = b[copy(d.x[:], b):]
+}
+
+func (h *sha3_224Hash) MarshalBinary() ([]byte, error) {
```

Comment:
The `x/crypto/sha3` package doesn't implement the marshaling functions. As doing so it quite tricky, I'll rather not do that. You can adapt the tests in `sha_test.go` to skip the checks that assume marshaling functions are implemented.